airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Base Asynchronous Operator for kicking off a long running
+operations and polling for completion with reschedule mode.
+"""
+
+from abc import abstractmethod
+from typing import Dict, List, Optional, Union
+
+from airflow.models import SkipMixin, TaskReschedule
+from airflow.models.xcom import XCOM_EXTERNAL_RESOURCE_ID_KEY
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.utils.decorators import apply_defaults
+
+
```

Comment:
> XCOM data appears to be cleared at start of each task. [See here](https://github.com/apache/airflow/blob/1a4c16432b8718189269e63c3afcd2709eed7379/airflow/models/taskinstance.py#L913).
> 
> So when task restarts after reschedule, we lose the resource id.

Ouch, that is quite a limitation for using XCom. Is there anywhere else to store task state? If not, some options:

1. Use `mode='poke'`, at the expense of using up a task slot for long-running tasks. Not ideal, but gets "correct" behaviour. It'd then be "atomic" rather than "async" behaviour.

1. Enhance `TaskInstance` to make clearing XCom data conditional by delegating it to the task/operator. There could be a new function like `BaseOperator.pre_execute_clear_state()` which can be overridden by implementers. `BaseOperator` is already aware of / coupled with `TaskInstance`, so I don't think we'd be breaking separation of concerns any more than it already is?

1. There might be enough justification to say that for rescheduled tasks (i.e. transition from `State.UP_FOR_RESCHEDULE` to `State.RUNNING`) then `TaskInstance` shouldn't clear XCom. The call to `run()` -> `_check_and_change_state_before_execution()` does know about this state change, but I see in the code that there are places which bypass the call to `run()` and go directly to `_run_raw_task()` (e.g. the CLI).

Reply:
Ref. `State.UP_FOR_RESCHEDULE`, this was added back in January by @seelmann as part of [AIRFLOW-2747](https://issues.apache.org/jira/browse/AIRFLOW-2747). Perhaps he has some thoughts on whether or not we can tighten the contract for rescheduled tasks to say that we retain XCom state? It's a fairly recent feature, so that might be okay, e.g. operators may not be depending on the current behaviour.

Reply:
It seems to me we definitely want to wipe XComs at the end of a DagRun as to not affect a re run of this dag for this execution date.

Reply:
What kind of consensus should we build on that kind of change for XCom in reschedule scenarios?

Reply:
> It seems to me we definitely want to wipe XComs at the end of a DagRun as to not affect a re run of this dag for this execution date.

if this is your concern, xcom can be cleared at start if dag is re-run. (indeed, that's how it works now, which is a problem for this PR)

probably makes sense to find some way to clear xcom when a task instance re-run but _not_ merely after a reschedule/poke.  then we can leave the xcom value in place at end of run, _and_ it is safe to re-run, and everybody's happy?

Reply:
@Fokko reviewed and approved PR #4408 where `State.UP_FOR_RESCHEDULE` was added, so he may also have an opinion? Given the purpose that this was added - for asynchronous sensors - and that it was added so relatively recently in 1.10.2, I can't think of a reason we shouldn't make the change to retain XCom on reschedule. Hopefully we can find a clean way to do that!

Reply:
If you need to persist state you may consider to add it to `task_reschedule` table? This was added in the initial reschedule PR https://github.com/apache/airflow/pull/3596/.

Reply:
Thank you for tagging me in this.

> Seems that some form of option 2 is the least risky way to get the desired outcome

I would not opt for option 2 since it will introduce two different behaviors (the updating of the xcom would be different, if it is async or not), and this might be confusing, and we're introducing code complexity by adding a branch.

My suggestion would be to not clear the xcom value, and when the operator finishes, we just overwrite the existing xcom value. WDYT?

> It seems to me we definitely want to wipe XComs at the end of a DagRun as to not affect a re run of this dag for this execution date.

We definitely don't want to do this(!) XComs can also be used across DAGs. So if you automatically clear the xcoms, they won't be available anymore.

> `State.UP_FOR_RESCHEDULE`

This was a state introduced for long running sensors. By definition take sensors a long time, for example, it can be that it will wait on another DAG to complete. This would be a simple lookup in the Airflow metastore.

`State.UP_FOR_RETRY` is the old state which means that the task could not complete. For example, when you are doing ETL, and you throw an exception, it will retry the task later on.

`State.UP_FOR_RESCHEDULE` means that it is waiting to be rescheduled again. There is a cooldown period before poking again. By making the transition from  `State.UP_FOR_RESCHEDULE` to `State.UP_FOR_RETRY` will let the scheduler know that it can be rerun again.

Two things:

- I'm not in favor of adding another table, such as `task_reschedule`. The xcom is already used to store state, so I think it woul be good to (mis)use it to store the state here as well.
- I'm also thinking about the class structure. Right now we have the BaseOperator -> BaseSensor -> BaseAsyncOperator, which feels a bit awkward. Ideally we would like to push the retry logic up in the tree.

I'll open up a PR, and see if we can upsert the xcom instead of clearing and inserting it.

Reply:
@seelmann and @Fokko thanks for jumping in and helping to build consensus!

> State.UP_FOR_RESCHEDULE means that it is waiting to be rescheduled again. There is a cooldown period before poking again. By making the transition from State.UP_FOR_RESCHEDULE to State.UP_FOR_RETRY will let the scheduler know that it can be rerun again.

Ref. the state changes, my interpretation of the [scheduler code](https://github.com/apache/airflow/blob/master/airflow/jobs/scheduler_job.py#L1557) is that the state change is from `UP_FOR_RESCHEDULE` to `SCHEDULED` to `QUEUED` to `RUNNING`? If that's the case then it does make my option 3 much more tricky, since it's entirely expected to want to retain the current behaviour in the `QUEUED` to `RUNNING` transition for other paths than `UP_FOR_RESCHEDULE`.

I do like the semantic separation between `UP_FOR_RETRY` and `UP_FOR_RESCHEDULE` because the latter implies that it's continuing work and state should be retained, whereas the former implies that it failed and the entire task should be retried.

> My suggestion would be to not clear the xcom value, and when the operator finishes, we just overwrite the existing xcom value. WDYT?
...
> I'll open up a PR, and see if we can upsert the xcom instead of clearing and inserting it.

Yes, upsert could be an interesting approach. It's not guaranteed to be clean if e.g. an operator were to have branches that cause different XCom keys to be output, though. I'm open to it as a compromise as it'll cover all practical cases that we're aware of today. Ref. backwards compatibility, how do we know that all operator instances will behave correctly if we stop clearing their state before starting?

> I'm also thinking about the class structure. Right now we have the BaseOperator -> BaseSensor -> BaseAsyncOperator, which feels a bit awkward. Ideally we would like to push the retry logic up in the tree.

Agreed, see [my comment](https://github.com/apache/airflow/pull/6210#discussion_r335271755) in the other thread about this.

Reply:
My pleasure. This is some cool functionality that a lot of users would benefit from. As @potiuk mentioned, let's see if we can make this happen :)

So the scheduler puts it to `SCHEDULED`, and then it the job will be pushed to the Celery queue (CeleryExecutor), Kubernetes (KubernetesExecutor) or a local queue (LocalExecutor) based on your executor. When there are resources available, in the case of Celery, a worker will pick it from the queue and start executing. When this happens, the worker will first put . the state from `QUEUED` to `RUNNING`.

The explicit clearing of the state goes back to Airflow 1.7. I think we can just update the state, instead of explicitly clearing it. I'm working on a PR here: https://github.com/apache/airflow/pull/6370

Since the xcom is a key-value structure, I think we should reserve some key for storing this state. Using xcom for this will also enable us to look at the state from the Airflow UI, this comes free.
The only downside for not clearing the xcom upfront, as far as I can see, is that some other task might still read old xcom data, in the case the task hasn't finished yet.

Reply:
@Fokko  thanks for jumping in. 
> Since the xcom is a key-value structure, I think we should reserve some key for storing this state. Using xcom for this will also enable us to look at the state from the Airflow UI, this comes free

@mik-laj had the same feedback so I added [`XCOM_EXTERNAL_RESOURCE_ID`](https://github.com/apache/airflow/pull/6210/files#diff-26749963c29f104c58fac199e5aa30c3R39). 

> I'm also thinking about the class structure. Right now we have the BaseOperator -> BaseSensor -> BaseAsyncOperator, which feels a bit awkward. Ideally we would like to push the retry logic up in the tree.

@Fokko @JonnyIncognito 
I'm assuming you mean push the reschedule logic up the tree. What would you think of `BaseOperator` > `BaseReschedulePokeOperator` > `BaseSensor` ?
We'd move the logic from `BaseSensorOperator`'s execute method into `BaseReschedulePokeOperator` (formerly known as BaseAsyncOperator) and then `BaseSensor` (formerly `BaseSensorOperator`) will become very simple w/ a `submit_request` of `pass`. It's a lot of code shuffling but I can make that change if we think that's a more sensible class structure. Do we see an argument for pushing reschedule logic into BaseOperator?

Reply:
I have added commit to illustrate the proposed class structure.
I'll put work into actually making all the tests work / etc. if you guys can confirm this is inline with what you're looking for.

Reply:
We should be careful with renaming classes. I think the async capabilities should be merged into the BaseOperator.

Reply:
I agree with trying to put the behaviour into BaseOperator. As it stands, the use-cases are:

1. Only operate (`BaseOperator` <- e.g. `EmrAddStepsOperator`)
1. Only sense (`BaseSensorOperator` <- e.g. `EmrStepSensor`)
1. Do both (to be determined)
1. Do none (`BaseOperator` <- e.g. `DummyOperator`)

Currently 1, 2 and 4 are implemented as variations of execute(); 2 is the odd one out in having a special class, as would 3 with the current proposal.

If BaseOperator had a default execute() that runs two phases, the other behaviours can be achieved by optionally implementing either of the phases.

It's a typical problem using class hierarchies for behaviours, that it's hard to mix and match. I was actually thinking about whether we could use traits to keep the behaviours cleanly separated into their own implementations. But it's probably overkill with so few.

Reply:
Something else to consider about a potential "multi phase" BaseOperator is that we already have `pre_execute()` + `execute()` + `post_execute()` and the order is governed by `TaskInstance._run_raw_task()` ([link](https://github.com/apache/airflow/blob/master/airflow/models/taskinstance.py#L857)). My understanding is that these hooks are always called even when tasks are rescheduled. I'm not entirely sure of the purpose, the docs simply say "... it’s mostly a hook for people deriving operators" and the only two places I can see in the source-code where they're used is: CloudSqlQueryOperator to ensure the db connection is created and torn down; and for Lineage support to ensure the lists of inlets and outlets are built and saved to Xcom (another kind of state retention!).

When considering extra phases, they'd be orthogonal to above and really some form of "execution phase N": the BaseOperator starts to behave a bit like a SubDag where the overall task status / state machine runs its course while what would have been sub-tasks in the SubDag now run as sequential execution phases in the same task. Instead of the sub-tasks being visible in the UI as DAG nodes, these execution phases are internalised into the BaseOperator.

Why would we want to do that? The justification is that these execution phases are tightly coupled and should be retried together (and in order) as some kind of atomic unit.

If we were to generalise this much, I think we should support being able to run through the execution phases synchronously (taking up a task slot) or asynchronously (using reschedule).

Sorry this is getting a bit philosophical but I'm trying to provide some guiding context about why we're doing this! I also see it as a potential future foundation for a mechanism to compose operators via the Airflow API.

Reply:
Back to the practical, can I propose:

- New BaseOperator hooks are implemented - something like `execute_phase_start()` and `execute_phase_final()` - that by default simply `raise NotImplementedError()` as per current `execute()` implementation

- New BaseOperator default `execute()` implementation that calls the two hooks in order. If the task was rescheduled, skip the first phase. I've seen code that detects this case in `BaseSensorOperator.execute()` ([link](https://github.com/apache/airflow/blob/master/airflow/sensors/base_sensor_operator.py#L104)).

- Existing overrides of `BaseOperator.execute()` should then behave as-is, i.e. the phase hooks are simply never called. The current `BaseSensorOperator` implementation would continue to work, minimising the amount of work needed in this PR; it could later be reimplemented to use the new phased execution.

- Any new operators implemented to combine an operation + sensor can override the new phase hooks, since they'd be multi-phase aware.

Allowing only two phases simplifies the implementation for now given the immediate requirements, but could even be extended in future for any number of phases if we wanted to store the current/next phase as part of the task state. I can't think of a good practical need right now.

What do you guys think, is this proposal (or something like it) workable?

Reply:
I can't see why we need another `execute_phase_{start,final}`. The pre-execute can be overridden by a custom implementation, but in practice, this hasn't been done much.

So the `TaskInstance._run_raw_task()` is what's being executed on the worker, and this will just actually run the whole task to the end.

> The justification is that these execution phases are tightly coupled and should be retried together (and in order) as some kind of atomic unit.

I really agree with that. We've seen situations when the DAG would start with creating a cluster, but this breaks the atomicity of the DAG. When the cluster is being killed (maybe because of idle timeout), the only way to retry the workflow, is by restarting the whole dag.

It is still not clear to me how the hierarchy will look like. I think we should move most of the reschedule logic into the BaseOperator. Maybe we can enrich the `context` which is being passed to the execute method with reschedule logic to let the implementation of the execute step know that you're rescheduling.